android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt

```diff
@@ -48,82 +45,80 @@ private const val RUDDER_SERVER_FILE_NAME_V1 = "RudderServerConfig"
  * Initialise session management for the user
  * Reinstate traits and external ids
  */
-class ReinstatePlugin : Plugin {
+private const val RUDDER_SERVER_FILE_NAME_V1 = "RudderServerConfig"
+
+internal class ReinstatePlugin : InfrastructurePlugin {
     private var _analytics: Analytics? = null
     private var sourceId: String? = null
-    private val isReinstated = AtomicBoolean(false)
-    private val stackedChainsTillReinstate = LinkedBlockingQueue<Plugin.Chain>()
-    override fun intercept(chain: Plugin.Chain): Message {
-        if(!isReinstated.get()){
-            stackedChainsTillReinstate.add(chain)
-            return chain.message()
-        }
-        processStackedMessages()
-        return chain.proceed(chain.message())
-    }
-
-    private fun processStackedMessages() {
-        if(!isReinstated.get()) return
-        synchronized(this) {
-            while (stackedChainsTillReinstate.isNotEmpty()) {
-                val stackedChain = stackedChainsTillReinstate.poll() ?: continue
-                stackedChain.proceed(stackedChain.message())
+    private val _isReinstated = AtomicBoolean(false)
+    private fun setReinstated(isReinstated: Boolean){
+        synchronized(_isReinstated) {
+            _isReinstated.set(isReinstated)
+            if (isReinstated){
+                _analytics?.applyInfrastructureClosure {
+                    if(this is DataUploadService)
+                        this.resume()
+                }
+            }else{
+                _analytics?.applyInfrastructureClosure {
+                    if(this is DataUploadService)
+                        this.pause()
+                }
             }
         }
     }
 
     override fun setup(analytics: Analytics) {
         _analytics = analytics
+        setReinstated(false)
     }
 
     override fun updateRudderServerConfig(config: RudderServerConfig) {
         val sourceId = config.source?.sourceId ?: return
         this.sourceId = sourceId
         reinstate()
-        processStackedMessages()
     }
 
     override fun updateConfiguration(configuration: Configuration) {
         reinstate()
     }
 
     private fun reinstate() {
-        if (isReinstated.get()) return
+        if (_isReinstated.get()) return
         val config = _analytics?.currentConfigurationAndroid ?: return
         if (isV2DataAvailable()) {
-            isReinstated.set(true)
-            reinstateV2FromCache(config)
+            setReinstated(true)
+            reinstateV2FromCache()
             return
         }
+        fillDefaults()
         if (!config.shouldVerifySdk){
-            defaults(config)
-            isReinstated.set(true)
+            setReinstated(true)
             return
         }
         val sourceId = this.sourceId ?: return
 
         migrateV1DataIfAvailable(config.application, sourceId, config)
```

Comment:
This `migrateV1DataIfAvailable(...)` code do not seems to be reachable. Please refactor the logic.